janusgraph-core/src/main/java/org/janusgraph/graphdb/util/SubqueryIterator.java

```diff
@@ -17,81 +17,236 @@
+                    continue;
+                int subLimit;
+                if (offsets[idx] > 0) {
+                    // Estimate the best subLimit based on previous round of query. It is a reasonable assumption
+                    // that no_of_total_results_needed : optimal_sublimit can be approximated by
+                    // no_of_last_intersected_results : last_sublimit.
+                    subLimit = results.size() > 0 ? limit * offsets[idx] / results.size() * 2
+                        : (int) Math.pow(offsets[idx], 1.5);
+                } else {
+                    subLimit = baseSubLimit;
+                }
+                subLimit = Math.min(Query.NO_LIMIT, subLimit);
+
+                executeQueries(queries, idx, subLimit, offsets, subResultToQueryMap, indexSerializer, tx);
+
+                if (offsets[idx] < subLimit) {
+                    resultsExhausted[idx] = true;
+                    resultsExhaustedCount++;
+                }
+            }
+
+            intersectResults(subResultToQueryMap, results, limit, queries.size());
+
+            scores = computeScores(resultsExhaustedCount, resultsExhausted, offsets, queries.size(), results.size(),
+                limit, subResultToQueryMap);
+
+        } while (resultsExhaustedCount < queries.size() && results.size() < limit);
+        elementIterator = results.stream().limit(limit).map(conversionFunction).map(r -> (JanusGraphElement) r).iterator();
+    }
+
+    private void executeQueries(List<JointIndexQuery.Subquery> queries, final int idx, final int subLimit,
+                                int[] offsets, Map<Object, List<Integer>> subResultToQueryMap,
+                                IndexSerializer indexSerializer, BackendTransaction tx) {
+        JointIndexQuery.Subquery subQuery = queries.get(idx).updateLimit(subLimit);
+        final List<Object> subQueryCache = indexCache.getIfPresent(subQuery);
+        if (subQueryCache != null) {
+            assert subQueryCache.size() >= offsets[idx];
+            for (int j = offsets[idx]; j < subQueryCache.size(); j++) {
+                subResultToQueryMap.computeIfAbsent(subQueryCache.get(j), k -> new ArrayList<>()).add(idx);
+            }
+            offsets[idx] = subQueryCache.size();
+        } else {
+            QueryProfiler profiler = subQuery.getProfiler();
+            QueryProfiler.startProfile(profiler, subQuery);
+            // TODO: leverage the scrolling capability of indexing backends ratherthan throw away old results
+            indexSerializer.query(subQuery, tx).skip(offsets[idx]).forEachOrdered(result -> {
+                offsets[idx]++;
+                subResultToQueryMap.computeIfAbsent(result, k -> new ArrayList<>()).add(idx);
+            });
             profiler.stopTimer();
+            profiler.setResultSize(offsets[idx]);
+            // TODO: should we put it into cache?
```

Comment:
I don't really have an opinion on that but I think it's at least worth a discussion.

Reply:
Yeah, there is one point I am unsure of, but I didn't have time to look into yet: if I put the first 5000 results of index query X into the cache, then when another index query X with limit 1000 comes, will that be effectively loaded from the cache? In other words, does the cache do exact match comparison or subset comparison based on the limit?

Reply:
I don't know anything about the cache implementation either, but maybe there's someone else here who does ;)